File: app/src/main/java/practicaltask/riseapps/com/practicaltask/ui/countryList/adapter/CountryListAdapter.java
package practicaltask.riseapps.com.practicaltask.ui.countryList.adapter;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import practicaltask.riseapps.com.practicaltask.R;
import practicaltask.riseapps.com.practicaltask.data.model.CountryDTO;
import practicaltask.riseapps.com.practicaltask.ui.base.Callback;

public class CountryListAdapter extends RecyclerView.Adapter<CountryListAdapter.ViewHolder> {

    private List<CountryDTO> mDataset;
    private Callback<CountryDTO> clickCallback;

    static class ViewHolder extends RecyclerView.ViewHolder {
        private View view;
        private TextView countryName;
        private TextView countryCapital;

        private ViewHolder(View view, TextView countryName, TextView countryCapital) {
            super(view);
            this.countryName = countryName;
            this.countryCapital = countryCapital;
            this.view = view;
        }
    }

    public CountryListAdapter(Callback<CountryDTO> clickCallback) {
        this.clickCallback = clickCallback;
    }

    @Override
    public CountryListAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.widget_country_element, parent, false);
        TextView name = view.findViewById(R.id.country_name);
        TextView capital = view.findViewById(R.id.country_capital);
        return new CountryListAdapter.ViewHolder(view, name, capital);
    }

    @Override
    public void onBindViewHolder(CountryListAdapter.ViewHolder holder, int position) {
        final CountryDTO country = mDataset.get(position);
        holder.countryName.setText(country.getName());
        holder.countryCapital.setText(country.getCapital());
    }

    @Override
    public int getItemCount() {
        return mDataset == null ? 0 : mDataset.size();
    }

    public void setmDataset(List<CountryDTO> mDataset) {
        this.mDataset = mDataset;
        notifyDataSetChanged();
    }
}

File: app/src/main/java/practicaltask/riseapps/com/practicaltask/ui/countryList/adapter/CountryListAdapter.kt
package practicaltask.riseapps.com.practicaltask.ui.countryList.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import practicaltask.riseapps.com.practicaltask.R
import practicaltask.riseapps.com.practicaltask.data.model.CountryDTO

class CountryListAdapter(private val clickCallback: (CountryDTO) -> Unit) : RecyclerView.Adapter<CountryListAdapter.ViewHolder>() {

    private var mDataset: List<CountryDTO>? = null

    inner class ViewHolder constructor(val view: View, val countryName: TextView, val countryCapital: TextView) : RecyclerView.ViewHolder(view)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CountryListAdapter.ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.widget_country_element, parent, false)
        val name = view.findViewById<TextView>(R.id.country_name)
        val capital = view.findViewById<TextView>(R.id.country_capital)
        return ViewHolder(view, name, capital)
    }

    override fun onBindViewHolder(holder: CountryListAdapter.ViewHolder, position: Int) {
        val country = mDataset!![position]
        holder.countryName.text = country.name
        holder.countryCapital.text = country.capital
        holder.view.setOnClickListener { v -> clickCallback(country) }
    }

    override fun getItemCount(): Int {
        return mDataset?.size ?: 0
    }

    fun setmDataset(mDataset: List<CountryDTO>) {
        this.mDataset = mDataset
        notifyDataSetChanged()
    }
}
